In [1]:
import pandas as pd

pd.set_option('display.max_columns', 150)

In [2]:
data = pd.read_csv("data/processed_data_2022-06-13.csv")
data.head()

sofifa_id                                         player_url  short_name  \
0     167575  https://sofifa.com/player/167575/anton-shunin/...   A. Shunin   
1     205211  https://sofifa.com/player/205211/milan-borjan/...   M. Borjan   
2     244700  https://sofifa.com/player/244700/yuriy-dyupin/...   Y. Dyupin   
3     199904  https://sofifa.com/player/199904/jesse-joronen...  J. Joronen   
4     240225  https://sofifa.com/player/240225/matvey-safono...  M. Safonov   

             long_name  overall  potential  value_eur  wage_eur         dob  \
0         Anton Shunin       76         76        NaN       NaN  1987-01-27   
1         Milan Borjan       76         76        NaN       NaN  1987-10-23   
2         Yuriy Dyupin       74         74        NaN       NaN  1988-03-17   
3  Jesse Pekka Joronen       72         72        NaN       NaN  1993-03-21   
4       Matvey Safonov       72         82        NaN       NaN  1999-02-25   

   weight_kg  height_cm  club_team_id club_position  club_jersey_number  \
0         88        191           NaN           NaN                 NaN   
1         84        196           NaN           NaN                 NaN   
2         84        187           NaN           NaN                 NaN   
3         89        197           NaN           NaN                 NaN   
4         82        192           NaN           NaN                 NaN   

   preferred_foot  weak_foot  skill_moves  international_reputation  \
0               0          3            1                         1   
1               1          2            1                         1   
2               1          3            1                         1   
3               1          3            1                         1   
4               1          2            1                         1   

   work_rate  body_type  release_clause_eur  pace  shooting  passing  \
0          0          0                 NaN   NaN       NaN      NaN   
1          0          0                 NaN   NaN       NaN      NaN   
2          0          0                 NaN   NaN       NaN      NaN   
3          0          0                 NaN   NaN       NaN      NaN   
4          0          0                 NaN   NaN       NaN      NaN   

   dribbling  defending  physic  attacking_crossing  attacking_finishing  \
0        NaN        NaN     NaN                  18                   11   
1        NaN        NaN     NaN                  15                   12   
2        NaN        NaN     NaN                  12                    6   
3        NaN        NaN     NaN                  15                   13   
4        NaN        NaN     NaN                  13                    7   

   attacking_heading_accuracy  skill_dribbling  skill_curve  \
0                          12               14           15   
1                          24               28           16   
2                          10                7           14   
3                          18               16           14   
4                          14               10           11   

   skill_fk_accuracy  skill_long_passing  skill_ball_control  \
0                 13                  30                  20   
1                 16                  48                  23   
2                 11                  22                  12   
3                 12                  25                  16   
4                 12                  26                  18   

   movement_acceleration  movement_sprint_speed  movement_agility  \
0                     50                     39                43   
1                     47                     40                40   
2                     20                     21                27   
3                     48                     45                50   
4                     19                     22                33   

   movement_reactions  movement_balance  power_shot_power  power_jumping  \
0                  71         

In [3]:
def count_na(df: pd.DataFrame) -> None:
    nan_sum = df.isna().sum()
    display(nan_sum[nan_sum > 0])

count_na(data)

value_eur                74
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

Analisando essa tabela com os valores nulos podemos perceber que alguns jogadores do dataset não possuem um time atual, isso explica
`wage_eur`, `club_team_id`, `club_position`, `club_jersey_number` terem a mesma quantidade de valores nulos (61),
porém `value_eur` deveria seguir esse mesmo padrão, assim podemos dizer que 13 (74 - 61) jogadores que estão ativos não possuem um valor,
portanto precisamos prever esses dados.

Também podemos perceber que temos 2132 jogadores não possuem `pace`, `shooting`, `passing`, `dribbling`, `defending`, `physic`. Vamos analisar esses jogadores

In [4]:
def percent_of_gk(df: pd.DataFrame) -> None:
    gk = df[df['GK'] == 1]
    display(gk.shape[0] / df.shape[0])


missing_main_attr_data = data[data.pace.isna()]

percent_of_gk(missing_main_attr_data)
missing_main_attr_data

1.0

sofifa_id                                         player_url  \
0        167575  https://sofifa.com/player/167575/anton-shunin/...   
1        205211  https://sofifa.com/player/205211/milan-borjan/...   
2        244700  https://sofifa.com/player/244700/yuriy-dyupin/...   
3        199904  https://sofifa.com/player/199904/jesse-joronen...   
4        240225  https://sofifa.com/player/240225/matvey-safono...   
...         ...                                                ...   
2127     257618  https://sofifa.com/player/257618/armin-gremsl/...   
2128     262424  https://sofifa.com/player/262424/sorin-mogosan...   
2129     262420  https://sofifa.com/player/262420/robert-popa/2...   
2130     210022  https://sofifa.com/player/210022/philip-sprint...   
2131     244417  https://sofifa.com/player/244417/julian-krahl/...   

       short_name            long_name  overall  potential  value_eur  \
0       A. Shunin         Anton Shunin       76         76        NaN   
1       M. Borjan         Milan Borjan       76         76        NaN   
2       Y. Dyupin         Yuriy Dyupin       74         74        NaN   
3      J. Joronen  Jesse Pekka Joronen       72         72        NaN   
4      M. Safonov       Matvey Safonov       72         82        NaN   
...           ...                  ...      ...        ...        ...   
2127    A. Gremsl         Armin Gremsl       63         66   475000.0   
2128  S. Mogoșanu       Sorin Mogoșanu       58         60   170000.0   
2129      R. Popa          Robert Popa       56         75   325000.0   
2130    P. Sprint        Philip Sprint       62         64   350000.0   
2131     J. Krahl         Julian Krahl       62         76   850000.0   

      wage_eur         dob  weight_kg  height_cm  club_team_id club_position  \
0          NaN  1987-01-27         88        191           NaN           NaN   
1          NaN  1987-10-23         84        196           NaN           NaN   
2          NaN  1988-03-17         84        187           NaN           NaN   
3          NaN  1993-03-21         89        197           NaN           NaN   
4          NaN  1999-02-25         82        192           NaN           NaN   
...        ...         ...        ...        ...           ...           ...   
2127    2000.0  1994-08-13         85        190      115716.0           RES   
2128    1000.0  1992-11-22         77        180      115716.0            GK   
2129     500.0  2003-03-05         83        190      115716.0           SUB   
2130     700.0  1993-06-27         98        196      115820.0            GK   
2131     500.0  2000-01-22         90        194      115820.0           SUB   

      club_jersey_number  preferred_foot  weak_foot  skill_moves  \
0                    NaN               0          3            1   
1                    NaN               1          2            1   
2                    NaN               1          3            1   
3                    NaN               1          3            1   
4                    NaN               1          2            1   
...                  ...             ...        ...          ...   
2127                94.0               0          2            1   
2128                 1.0               1          1            1   
2129                31.0               1          2            1   
2130                 1.0               1          2            1   
2131                13.0               1          3            1   

      international_reputation  work_rate  body_type  release_clause_eur  \
0                            1          0          0                 NaN   
1                            1          0          0                 NaN   
2                            1          0          0                 NaN   
3                            1          0          0                 NaN   
4                            1          0          0                 NaN   
...                        ...        ...        ...  

Como vimos que os goleiros não tem os atributos listados acima, mas possuem um overall em cada posição, seria interessante prever esses atributos.

In [5]:
missing_gk_speed_data = data[data.goalkeeping_speed.isna()]
sum_of_missing_data = missing_gk_speed_data.shape[0] + missing_main_attr_data.shape[0]
print(f"Jogadores sem gk speed + jogadores sem atributos primários: {sum_of_missing_data}")
print(f"Numero de jogadores: {data.shape[0]}")

Jogadores sem gk speed + jogadores sem atributos primários: 19239
Numero de jogadores: 19239


Assim podemos ver que todos os jogadores que não são goleiros não possuem `goalkeeping_speed`

## Prevendo valor dos jogadores
Nessa etapa vamos criar um modelo para prever o valor dos 13 jogadores que não possuem valor

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
import numpy as np

def predict_and_populate(df_to_populate, df_to_train, model, column_to_predict, round_=None) -> pd.DataFrame:
    # Separating the target variable
    target = df_to_train[column_to_predict]
    features = df_to_train.drop(column_to_predict, axis=1)

    # Poluting the missing features
    mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
    mean_imputer.fit(features)
    imputed_features = mean_imputer.transform(features)

    # Train test split
    features_train, features_test, target_train, target_test = train_test_split(imputed_features, target, test_size=0.2)

    # Training the model
    trained_model = model.fit(features_train, target_train)

    # Calculatin model score
    model_score = trained_model.score(features_test, target_test)
    print(f"Score do modelo: {model_score:.4f}")

    # Populating the missing values
    feature_to_populate = mean_imputer.transform(df_to_populate.drop(column_to_predict, axis=1))
    predicted_values = trained_model.predict(feature_to_populate)

    if round_ is not None:
        predicted_values = predicted_values.round(round_)

    df_to_populate[column_to_predict] = predicted_values

    return df_to_populate

In [7]:
players_with_club_and_no_value = data.query("not club_team_id.isna() and value_eur.isna()").copy()
data_without_players_with_club_and_no_value = data.drop(players_with_club_and_no_value.index)

data_to_train = data_without_players_with_club_and_no_value.query("not value_eur.isna()").copy()
columns_to_drop = ["sofifa_id", "player_url", "short_name", "long_name", "dob", "club_position"]


data_to_train = data_to_train.drop(columns=columns_to_drop)
data_to_fill = players_with_club_and_no_value.drop(columns=columns_to_drop)

In [8]:
# Normalizando os dados
# from sklearn.preprocessing import StandardScaler
#
# standerd_scaler =  StandardScaler()
# standardized_features = standerd_scaler.fit_transform(imputed_features)
# standardized_features

## Comparando modelos diferentes

In [9]:
from sklearn.dummy import DummyRegressor

dummy = DummyRegressor()
predict_and_populate(data_to_fill, data_to_train, dummy, "value_eur");

Score do modelo: -0.0001


In [10]:
from sklearn.linear_model import LinearRegression

linear_regression = LinearRegression()
predict_and_populate(data_to_fill, data_to_train, linear_regression, "value_eur");

Score do modelo: 0.9752


In [11]:
from sklearn.ensemble import RandomForestRegressor

random_forest = RandomForestRegressor(random_state=0, n_jobs=-1, oob_score=True)
players_with_club_and_no_value = predict_and_populate(data_to_fill, data_to_train, random_forest, "value_eur");

Score do modelo: 0.9923


### Preenchendo os dados a partir do modelo criado

In [12]:
data.loc[players_with_club_and_no_value.index, "value_eur"] = players_with_club_and_no_value.value_eur
data.loc[players_with_club_and_no_value.index]

sofifa_id                                         player_url  \
11929     115909  https://sofifa.com/player/115909/ruben-castro-...   
11987     124344  https://sofifa.com/player/124344/ibrahim-oztur...   
12264     153066  https://sofifa.com/player/153066/lucas-licht/2...   
13149     152912  https://sofifa.com/player/152912/jose-sand/220002   
13177     110381  https://sofifa.com/player/110381/maximiliano-r...   
14103     262371  https://sofifa.com/player/262371/craig-dundas/...   
14611     140222  https://sofifa.com/player/140222/paulo-da-silv...   
15125     138703  https://sofifa.com/player/138703/andres-daless...   
16065     158293  https://sofifa.com/player/158293/zhi-zheng/220002   
16498       2702  https://sofifa.com/player/2702/kevin-ellison/2...   
16814      29118  https://sofifa.com/player/29118/shinji-ono/220002   
18159      29552  https://sofifa.com/player/29552/shunsuke-nakam...   
18166     254704  https://sofifa.com/player/254704/kazuyoshi-miu...   

            short_name                     long_name  overall  potential  \
11929     Rubén Castro           Rubén Castro Martín       69         69   
11987        İ. Öztürk                İbrahim Öztürk       67         67   
12264         L. Licht            Lucas Matías Licht       69         69   
13149          J. Sand             José Gustavo Sand       75         75   
13177     M. Rodríguez   Maximiliano Rubén Rodríguez       73         73   
14103        C. Dundas                  Craig Dundas       56         56   
14611      P. Da Silva  Paulo César Da Silva Barrios       67         67   
15125  A. D'Alessandro   Andrés Nicolás D'Alessandro       75         75   
16065        Zheng Zhi                            郑智       67         67   
16498       K. Ellison           Kevin Keith Ellison       58         58   
16814           S. Ono                         小野 伸二       58         58   
18159      S. Nakamura                         中村 俊輔       63         63   
18166         K. Miura                         三浦 知良       59         59   

       value_eur  wage_eur         dob  weight_kg  height_cm  club_team_id  \
11929   905000.0    3000.0  1981-06-27         68        169      100851.0   
11987   801750.0    3000.0  1981-06-21         80        186      101006.0   
12264   921000.0    5000.0  1981-04-06         72        174      101084.0   
13149  3751000.0   13000.0  1980-07-17         79        182      110395.0   
13177  2380750.0   10000.0  1981-01-02         74        173      110396.0   
14103   190150.0    1000.0  1981-02-16         89        188      110799.0   
14611   796750.0     500.0  1980-02-01         76        180      111007.0   
15125  3986000.0     900.0  1981-04-15         69        174      111325.0   
16065   841000.0    5000.0  1980-08-20         75        180      111839.0   
16498   281250.0    2000.0  1979-02-23         76        185      112254.0   
16814   308150.0     800.0  1979-09-27         76        175      112444.0   
18159   634250.0     800.0  1978-06-24         71        178      113197.0   
18166   344250.0     700.0  1967-02-26         72        177      113197.0   

      club_position  club_jersey_number  preferred_foot  weak_foot  \
11929            ST                 7.0               1          3   
11987            CB                38.0               1          2   
12264           SUB                25.0               0          4   
13149            LS                 9.0               1          4   
13177           SUB                11.0               1          4   
14103           RES                14.0               1          3   
14611           LCB                 3.0               1          3   
15125           SUB                10.0               0          3   
16065           LCM                10.0               1          4   
16498           SUB                22.0               0          3   
16814           SUB                44.0               1          4   
18159         

In [13]:
count_na(data)

value_eur                61
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

In [14]:
# Removendo jogadores não ativos
data = data.query("not club_team_id.isna()").copy()

count_na(data)

release_clause_eur     1115
pace                   2124
shooting               2124
passing                2124
dribbling              2124
defending              2124
physic                 2124
goalkeeping_speed     17054
dtype: int64

In [15]:
# TODO acidionar .joblib ou .pickle do modelo

# Prevendo pace, shooting, passing, dribbling, defending, physic dos goleiros

In [16]:
gk_data = data.query("GK == 1").copy()
data_without_gk = data.query("GK == 0").copy()

columns_to_predict = ["pace", "shooting", "passing", "dribbling", "defending", "physic"]

for column_to_predict in columns_to_predict:

    columns_to_drop_with_other_atributes = [col for col in columns_to_predict if col != column_to_predict]
    columns_to_drop_with_other_atributes.extend(columns_to_drop)

    data_to_train = data_without_gk.drop(columns=columns_to_drop_with_other_atributes)
    data_to_predict = gk_data.drop(columns=columns_to_drop_with_other_atributes)

    linear_regression = LinearRegression()
    print(f"Prevendo {column_to_predict}")
    filled_data = predict_and_populate(data_to_predict, data_to_train, linear_regression, column_to_predict, 0)
    print()

    data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]

# data.loc[gk_data.index]

Prevendo pace
Score do modelo: 0.9993

Prevendo shooting
Score do modelo: 0.9989

Prevendo passing
Score do modelo: 0.9961

Prevendo dribbling
Score do modelo: 0.9991

Prevendo defending
Score do modelo: 0.9997

Prevendo physic
Score do modelo: 0.9991



In [17]:
count_na(data)

release_clause_eur     1115
goalkeeping_speed     17054
dtype: int64

## Prevendo goalkeeping_speed dos jogadores

In [18]:
players_data = data.query("GK == 0").copy()
players_data

sofifa_id                                         player_url  \
2185      188567  https://sofifa.com/player/188567/pierre-emeric...   
2186      209989  https://sofifa.com/player/209989/thomas-partey...   
2187      193301  https://sofifa.com/player/193301/alexandre-lac...   
2188      222665  https://sofifa.com/player/222665/martin-odegaa...   
2189      226110  https://sofifa.com/player/226110/nicolas-pepe/...   
...          ...                                                ...   
19234     262593  https://sofifa.com/player/262593/moritz-seiffe...   
19235     262614  https://sofifa.com/player/262614/till-muschkow...   
19236     262755  https://sofifa.com/player/262755/shalva-ogbaid...   
19237     262756  https://sofifa.com/player/262756/matteo-gumane...   
19238     263236  https://sofifa.com/player/263236/yazid-heimur/...   

           short_name                                    long_name  overall  \
2185    P. Aubameyang  Pierre-Emerick Emiliano François Aubameyang       85   
2186        T. Partey                                Thomas Partey       83   
2187     A. Lacazette                          Alexandre Lacazette       82   
2188      M. Ødegaard                              Martin Ødegaard       82   
2189          N. Pépé                                 Nicolas Pépé       81   
...               ...                                          ...      ...   
19234     M. Seiffert                              Moritz Seiffert       58   
19235  T. Muschkowski                             Till Muschkowski       55   
19236     S. Ogbaidze                              Shalva Ogbaidze       55   
19237      M. Gumaneh                               Matteo Gumaneh       55   
19238       Y. Heimur                                 Yazid Heimur       55   

       potential   value_eur  wage_eur         dob  weight_kg  height_cm  \
2185          85  43500000.0  145000.0  1989-06-18         80        187   
2186          83  36500000.0  115000.0  1993-06-13         77        185   
2187          82  30000000.0  110000.0  1991-05-28         73        175   
2188          88  49000000.0   89000.0  1998-12-17         68        178   
2189          82  30500000.0   95000.0  1995-05-29         73        183   
...          ...         ...       ...         ...        ...        ...   
19234         66    450000.0     600.0  2000-11-04         74        180   
19235         65    275000.0     550.0  2001-05-06         72        182   
19236         65    275000.0     500.0  2002-08-01         75        186   
19237         67    300000.0     500.0  2002-09-05         76        188   
19238         66    275000.0     500.0  2002-11-18         74        186   

       club_team_id club_position  club_jersey_number  preferred_foot  \
2185            1.0            ST                14.0               1   
2186            1.0           RDM                 5.0               1   
2187            1.0           SUB                 9.0               1   
2188            1.0           CAM                 8.0               0   
2189            1.0            RM                19.0               0   
...             ...           ...                 ...             ...   
19234      115820.0           SUB                23.0               0   
19235      115820.0           SUB                18.0               1   
19236      115820.0           RES                28.0               1   
19237      115820.0           SUB                21.0               1   
19238      115820.0           RES                15.0               1   

       weak_foot  skill_moves  international_reputation  work_rate  body_type  \
2185           4            4                         3          0          0   
2186           3            3                         2          0          0   
2187           4            4                         3          1          0   
2188           2            5                         3          0          0   
2189      

In [19]:
data_to_predict = players_data.drop(columns=columns_to_drop)
data_to_train = gk_data.drop(columns=columns_to_drop)

column_to_predict = "goalkeeping_speed"

linear_regression = LinearRegression()
filled_data = predict_and_populate(data_to_predict, data_to_train, linear_regression, column_to_predict, 0)

data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]

Score do modelo: 0.9993


In [20]:
count_na(data)

release_clause_eur    1115
dtype: int64

In [21]:
import os
def get_project_root_path():
    project_directory = "fifa22-players-analysis"
    separated_path = os.getcwd().split(os.sep)
    project_directory_index = separated_path.index(project_directory) + 1
    return os.sep.join(separated_path[0:project_directory_index])

In [22]:
from datetime import date
data.drop(columns=["release_clause_eur"], inplace=True)

filled_data_save_path = os.sep.join([get_project_root_path(), "preprocessing", "data", f"filled_data_{date.today()}.csv"])
data.to_csv(filled_data_save_path, index=False)
filled_data_save_path

'/home/gustavo/fifa22-players-analysis/preprocessing/data/filled_data_2022-06-14.csv'

In [23]:
count_na(data)

Series([], dtype: int64)